Add Git variables migration documentation (#1641)
OctopusDeploy/docs: docs/projects/version-control/converting/migrating-variables.md

---
title: Migrating variables to Git
description: Variables needs to be manually migrated to Git for existing projects
position: 110
hideInThisSection: true
---

> Support for storing non-senstive variables in Git will be available to everyone in an upcoming version of Octopus. All newly configured Git projects will automatically have variables migrated when converting the project to Git - for existing Git projects, you will need to migrate these manually.

Reviewer: If we know the version that will support it, we should include that to avoid ambiguity (in my opinion)

```suggestion
Support for storing non-sensitive variables in Git will be available to everyone in Octopus **2022.3**. All newly configured Git projects will automatically have variables migrated when converting the project to Git. For existing Git projects, you will need to migrate these manually.
```

Author: My plan with this is to be intentionally vague (for now), and put in an update with the specific versions once we've got the toggle enabled by default for everyone. I don't want to just put `2022.3` in there, because there are a bunch of cloud `2022.3` versions that do not support Git variables by default and may cause confusion.

Once we've got it locked down to the specific `2022.3` and `2022.4` versions that have it, I plan to update this.

Do we have a typical approach for things like this? We want to have the documentation out before we enable the feature for everyone, but we don't know the specific version until we do.

Reviewer: That makes sense. Being vague is fine :smile:. There have been two approaches in the past:

1. Specify the exact version e.g. 2022.3.12345 it's available from
2. Specifying wording that doesn't commit to a version and then adding in the overall version e.g. 2022.3 when it's GA to both Cloud and Server (on-prem)

Both work :+1:

:::hint

> We recommend migrating your variables to Git as soon as practical. We will continue to support Git projects with un-migrated variables for now, but we will remove support for this at some point in the future.

Reviewer: ```suggestion
We recommend migrating your variables to Git as soon as practical. We will continue to support Git projects with variables that haven't been migrated for now, but we will remove support for this at some point in the future.
```

:::

> This document covers the process of migrating variables to Git.
> 
> ## Migrating variables to Git
> 
> You will need to migrate the variables for each existing Git project separately.

Reviewer: ```suggestion
This document covers the process of migrating variables to Git. You will need to migrate the variables for each existing Git project separately.
```

> ### Preparing to migrate
> 
> #### Check that all secrets are marked as sensitive

Reviewer: ```suggestion
## Preparing to migrate

### Check that all secrets are marked as sensitive
```

> All non-sensitive values will be written to your Git repository in plain text during the migration. Before migrating, make sure that you do not have any secret values that are not stored as sensitive variables.

Reviewer: ```suggestion
All non-sensitive values will be written to your Git repository in plain text during the migration. Before migrating, ensure that you do not have any secret values that are not stored as sensitive variables.
```

:::warning

> If you can view a value on the on the variables page (and it's not shown as a password field) or retrieve the value it from the API, it's not marked as sensitive and will be written to the `variables.ocl` file.

Reviewer: ```suggestion
If you can view a value on the variables page (and it's not shown as a password field) or retrieve the value from the API, it's not marked as sensitive and will be written to the `variables.ocl` file.
```

:::

> #### Coordinate with your team

Reviewer: ```suggestion
### Coordinate with your team
```

> The migration process will only write the variables to a single branch. After migrating you will need to merge this change to your other branches to continue using those branches in Octopus. If a branch does not have a `variables.ocl` file after the migration, you will get an error when trying to read.

Reviewer: ```suggestion
The migration process will only write the variables to a single branch. After migrating, you will need to merge this change to your other branches to continue using those branches in Octopus. If a branch does not have a `variables.ocl` file after the migration, you will get an error when trying to read.
```

Reviewer 2: Can borrow some words that I wrote for the warning dialog:
> Any existing Git tags or branches that do not have the migration commit will not be able to create releases, run runbooks, edit deployment processes, update variables etc.

Author: I've reworked this to remove the vague "you will get an error", and add some examples of when it might fail. I don't want to call out specific error messages, because they're likely to change, there are a few possible error messages, and a few possible causes of those error messages.

This is what I have now:

```
### Coordinate with your team

After you have run the migration, you will be unable to read variables from any Git reference (branch, tag, or commit) that does not contain a valid `variables.ocl` file. This will cause failues when viewing and editing variables, creating releases, running runbooks using those references until they're updated.

As the migration process only writes the variables to a single branch, you will need to merge that migration branch with your other branches to continue using them in Octopus. If this may disrupt your team, plan ahead and find a suitable time to migrate and merge.
```

Reviewer: LGTM. You _could_ remove the As and start with `The` e.g.:

> The migration process only writes the variables to a single branch. To continue using the variables in any other branches, you will need to merge that migration branch with those branches. If this may disrupt your team, plan ahead and find a suitable time to migrate and merge.

> If this may be disruptive to your team, plan ahead and find a suitable time to migrate and merge.

Reviewer: ```suggestion
If this may disrupt your team, plan ahead and find a suitable time to migrate and merge.
```

> ### Running the migration

Reviewer: ```suggestion
## Running the migration
```

The migration process is simple, and we've built a helpful wizard to guide you through it. If this feature is available in your instance and you have a project that does not yet have variables in Git, you will see this banner on the project variables page.

> ![ALT TEXT TODO](git-variables-migrator-banner.png "width=400")

Reviewer: ```suggestion
![Config as Code variable migration banner](git-variables-migrator-banner.png "width=400")
```

Clicking the **Migrate Variables To Git** button will open the migation wizard.

> #### Step 1: Getting started

Reviewer: ```suggestion
### Step 1: Getting started
```

The first page is just informational. After opening the dialog, Octopus will validate that the project can migrate variables to Git, then you can continue.

> ![ALT TEXT TODO](git-variables-migrator-page-1.png "width=400")

Reviewer: ```suggestion
![Config as Code variable migration wizard - Getting started](git-variables-migrator-page-1.png "width=400")
```

> #### Step 2: Select a branch

Reviewer: ```suggestion
### Step 2: Select a branch
```

On this page you select a branch to migrate the variables to.

Always use the default branch if possible. The default branch is the first branch shown when new users view the project, and Octopus only snapshots variables from the default branch for Runbooks. You always want to have a valid variables file on the default branch, so migrate directly there _if you can_.

> ![ALT TEXT TODO](git-variables-migrator-page-2-existing.png "width=400")

Reviewer: ```suggestion
![Config as Code variable migration wizard - Select branch](git-variables-migrator-page-2-existing.png "width=400")
```

> If you are unable migrate to the default branch (for example, it's protected), you can select any other branch or get Octopus to create a new branch from the default when migrating. _Once the migration is finished, merge the variables to the default branch as soon as possible._

Reviewer: ```suggestion
If you are unable to migrate to the default branch (for example, it's protected), you can select any other branch or get Octopus to create a new branch from the default when migrating. _Once the migration is finished, merge the variables to the default branch as soon as possible._
```

> #### Step 3: Review & migrate

Reviewer: ```suggestion
### Step 3: Review & migrate
```

After selecting the branch, you can view the migration summary and start the migration. The migration is typically completed in a few seconds.

> ![ALT TEXT TODO](git-variables-migrator-page-3.png "width=400")

Reviewer: ```suggestion
![Config as Code variable migration wizard - Summary](git-variables-migrator-page-3.png "width=400")
```

Once the migration is complete, click **Done**. The project will reload, and your variables will now be in Git.

> ### Next steps
> #### Merge variables to the default branch

Reviewer: ```suggestion
## Next steps
### Merge variables to the default branch
```

If you migrated the variables to any branch other than the default, be sure merge the variables to the default branch immediately. The variables are now just like any other file in your Git respository, so follow your typical Git workflows to get this merged.

> #### Update any other branches

Reviewer: ```suggestion
### Update any other branches
```

> If you're a frequent user of branching within Octopus, you will want to make sure you get the variables to all of your active branches as quickly as possible as well.

Reviewer: ```suggestion
If you're a frequent user of branching within Octopus, you will want to ensure you get the variables to all of your active branches as quickly as possible.
```

> Once the variables have been merged to the default branch, you can update your branches as you would with any other files in your repository - pull the latest changes into your branch and you'll be good to go.

Reviewer: ```suggestion
Once the variables have been merged to the default branch, you can update your branches as you would with any other files in your repository - pull the latest changes into your branch, and you'll be good to go.
```
